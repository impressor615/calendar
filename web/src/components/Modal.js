import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Button from 'components/Button';

export const ModalHeader = ({ children, onToggle }) => (
  <div className="modal-header">
    <div>{children}</div>
    <Button onClick={onToggle}>&times;</Button>
  </div>
);

ModalHeader.propTypes = {
  children: PropTypes.node.isRequired,
  onToggle: PropTypes.func.isRequired,
};

export const ModalBody = ({ children }) => (
  <div className="modal-body">
    {children}
  </div>
);

ModalBody.propTypes = {
  children: PropTypes.node.isRequired,
};

export const ModalFooter = ({ children }) => (
  <div className="modal-footer">
    { children }
  </div>
);

ModalFooter.propTypes = {
  children: PropTypes.node.isRequired,
};

export const ModalDialog = ({ children, ...rest }) => (
  <div className={classnames('modal-dialog', rest.className)}>
    { children }
  </div>
);

ModalDialog.propTypes = {
  children: PropTypes.node.isRequired,
};

const Modal = ({ isOpen, children, backdrop, onToggle, ...rest }) => (
  isOpen ? (
    ReactDOM.createPortal(
      <div
        {...rest}
        className={classnames('modal', rest.className)}
      >
        {children}
        { backdrop ? <div className="modal-backdrop" onClick={onToggle} /> : null }
      </div>,
      document.getElementById('modal-root'),
    )
  ) : null
);

Modal.defaultProps = {
  backdrop: true,
  onToggle: () => {},
};

Modal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired,
  onToggle: PropTypes.func,
  backdrop: PropTypes.bool,
};

export default Modal;
