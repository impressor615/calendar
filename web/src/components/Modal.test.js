import React from 'react';
import Modal, { ModalDialog, ModalHeader, ModalBody, ModalFooter} from 'components/Modal';


describe('<Modal />', () => {
  const modalRoot = document.createElement('div');
  modalRoot.setAttribute('id', 'modal-root');
  const body = document.querySelector('body');
  body.appendChild(modalRoot);

  const onToggle = jest.fn();
  const props = {
    isOpen: true,
  };
  let wrapper;
  const component = (
    <Modal {...props}>
      <ModalDialog>
        <ModalHeader onToggle={onToggle}>
          일정 추가 및 수정
        </ModalHeader>
        <ModalBody>
          ModalBody
        </ModalBody>
        <ModalFooter>
          <button>취소</button> 
          <button>저장</button>
        </ModalFooter>
      </ModalDialog>
    </Modal>
  );

  afterEach(() => {
    wrapper.unmount();
  })

  it('should be rendered properly', () => {
    wrapper = mount(component);

    expect(wrapper.find(ModalDialog)).toHaveLength(1);
    expect(wrapper.find(ModalHeader)).toHaveLength(1);
    expect(wrapper.find(ModalBody)).toHaveLength(1);
    expect(wrapper.find(ModalFooter)).toHaveLength(1);
  })

  it('shoud be matched with the snapshot', () => {
    wrapper = mount(component);
    expect(wrapper).toMatchSnapshot();
  })
});
