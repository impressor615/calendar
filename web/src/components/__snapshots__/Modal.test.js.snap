// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> shoud be matched with the snapshot 1`] = `
<Modal
  backdrop={true}
  isOpen={true}
  onToggle={[Function]}
>
  <Portal
    containerInfo={
      <div
        id="modal-root"
      >
        <div
          class="modal"
        >
          <div
            class="modal-dialog"
          >
            <div
              class="modal-header"
            >
              <div>
                일정 추가 및 수정
              </div>
              <button
                class="btn"
                type="button"
              >
                ×
              </button>
            </div>
            <div
              class="modal-body"
            >
              ModalBody
            </div>
            <div
              class="modal-footer"
            >
              <button>
                취소
              </button>
              <button>
                저장
              </button>
            </div>
          </div>
          <div
            class="modal-backdrop"
          />
        </div>
      </div>
    }
  >
    <div
      className="modal"
    >
      <ModalDialog>
        <div
          className="modal-dialog"
        >
          <ModalHeader
            onToggle={[MockFunction]}
          >
            <div
              className="modal-header"
            >
              <div>
                일정 추가 및 수정
              </div>
              <button
                className="btn"
                onClick={[MockFunction]}
                type="button"
              >
                ×
              </button>
            </div>
          </ModalHeader>
          <ModalBody>
            <div
              className="modal-body"
            >
              ModalBody
            </div>
          </ModalBody>
          <ModalFooter>
            <div
              className="modal-footer"
            >
              <button>
                취소
              </button>
              <button>
                저장
              </button>
            </div>
          </ModalFooter>
        </div>
      </ModalDialog>
      <div
        className="modal-backdrop"
        onClick={[Function]}
      />
    </div>
  </Portal>
</Modal>
`;
